pkg/providers/vsphere/upgrade/virtualmachine/vm_schema_upgrade.go

```diff
@@ -54,43 +59,117 @@ func ReconcileSchemaUpgrade(
+	var (
+		logger   = pkglog.FromContextOrDefault(ctx)
+		features = pkgcfg.FromContext(ctx).Features
+	)
+
 	logger.V(4).Info("Reconciling schema upgrade for VM")
 
+	if vmopv1util.IsVirtualMachineSchemaUpgraded(ctx, *vm) {
+
+		if features.AllDisksArePVCs || features.VMSharedDisks {
+			// Ensure the backfill condition is marked ready since the backfill
+			// reconcile will not be called again since it's been marked as
+			// upgraded.
+			pkgcond.MarkTrue(vm, vmconfunmanagedvolsfill.Condition)
+		}
+
+		logger.V(4).Info(
+			"Skipping reconciliation of schema upgrade for VM" +
+				" that is already upgraded")
+		return nil
+	}
+
 	var (
+		wasModified        bool
+		wasFeatureModified bool
+
 		curBuildVersion  = pkgcfg.FromContext(ctx).BuildVersion
 		curSchemaVersion = vmopv1.GroupVersion.Version
 
-		vmBuildVersion  = vm.Annotations[pkgconst.UpgradedToBuildVersionAnnotationKey]
-		vmSchemaVersion = vm.Annotations[pkgconst.UpgradedToSchemaVersionAnnotationKey]
+		vmBuildVersion   = vm.Annotations[pkgconst.UpgradedToBuildVersionAnnotationKey]
+		vmSchemaVersion  = vm.Annotations[pkgconst.UpgradedToSchemaVersionAnnotationKey]
+		vmFeatureVersion = vmopv1util.ParseFeatureVersion(
+			vm.Annotations[pkgconst.UpgradedToFeatureVersionAnnotationKey])
 	)
 
-	if vmBuildVersion == curBuildVersion &&
-		vmSchemaVersion == curSchemaVersion {
+	if vmBuildVersion != curBuildVersion {
+		vm.SetAnnotation(
+			pkgconst.UpgradedToBuildVersionAnnotationKey,
+			curBuildVersion)
+		wasModified = true
+	}
+
+	if vmSchemaVersion != curSchemaVersion {
+		vm.SetAnnotation(
+			pkgconst.UpgradedToSchemaVersionAnnotationKey,
+			curSchemaVersion)
+		wasModified = true
+	}
+
+	if f := vmopv1util.FeatureVersionBase; !vmFeatureVersion.Has(f) {
+		reconcileBIOSUUID(ctx, vm, moVM)
+		reconcileInstanceUUID(ctx, vm, moVM)
+		reconcileCloudInitInstanceUUID(ctx, vm, moVM)
 
-		logger.V(4).Info("Skipping reconciliation of schema upgrade for VM" +
-			" that is already upgraded")
-		return nil
+		wasFeatureModified = true
+		vmFeatureVersion.Set(f)
 	}
 
-	reconcileBIOSUUID(ctx, vm, moVM)
-	reconcileInstanceUUID(ctx, vm, moVM)
-	reconcileCloudInitInstanceUUID(ctx, vm, moVM)
-	reconcileControllers(ctx, vm, moVM)
-	reconcileDevices(ctx, vm, moVM, k8sClient)
+	if features.AllDisksArePVCs || features.VMSharedDisks {
+		if f := vmopv1util.FeatureVersionAllDisksArePVCs; !vmFeatureVersion.Has(f) {
+			if err := vmconfunmanagedvolsfill.Reconcile(
+				ctx,
+				nil,
+				nil,
+				vm,
+				moVM,
+				nil); err != nil {
 
-	// Indicate the VM has been upgraded.
-	if vm.Annotations == nil {
-		vm.Annotations = map[string]string{}
+				if !errors.Is(err, vmconfunmanagedvolsfill.ErrPendingBackfill) {
+					return fmt.Errorf(
+						"unexpected unmanaged disk backfill error: %w", err)
+				}
+			}
+
+			wasFeatureModified = true
+			vmFeatureVersion.Set(f)
+		}
+	}
+
+	if features.VMSharedDisks {
+		if f := vmopv1util.FeatureVersionVMSharedDisks; !vmFeatureVersion.Has(f) {
```

Comment:
Boot disks need to be backfilled for VMSharedDisks, that's why `vmconfunmanagedvolsfill.Reconcile` was called before for both features

Reply:
Yes, which still happens. Am I missing something obvious?